src/inmanta/server/services/orchestrationservice.py

```diff
@@ -468,117 +395,155 @@ async def delete_version(self, env: data.Environment, version_id: int) -> Apiret
+
+        rids = set(rid_to_resource.keys())
+
+        # Sanity checks
+        superfluous_ids = set(resource_sets.keys()) - rids
         if superfluous_ids:
             raise BadRequest(
                 "The following resource ids provided in the resource_sets parameter are not present "
                 f"in the resources list: {', '.join(superfluous_ids)}"
             )
-        requires_ids = set(provides_tree.keys())
-        if not requires_ids.issubset(resource_ids):
+        if not all_requires.issubset(rids):
             raise BadRequest(
                 "The model should have a dependency graph that is closed and no dangling dependencies:"
-                f" {requires_ids-resource_ids}"
+                f" {all_requires - rids}"
             )
-        # hook up all CADs
-        for f, t in cross_agent_dep:
-            res_obj = rv_dict[t.resource_str()]
-            res_obj.provides.append(f.resource_id)
+
+        return rid_to_resource
+
+    def _get_skipped_for_undeployable(
+        self, resources: abc.Sequence[data.Resource], undeployable_ids: abc.Sequence[ResourceIdStr]
+    ) -> abc.Sequence[ResourceIdStr]:
+        """
+        Return the resources that are skipped_for_undeployable given the full set of resources and
+        the resource ids of the resources that are undeployable.
+
+        :param resources: All resources in the model.
+        :param undeployable_ids: The ids of the resource that are undeployable.
+        """
+        # Build up provides tree
+        provides_tree: Dict[ResourceIdStr, List[ResourceIdStr]] = defaultdict(lambda: [])
+        for r in resources:
+            if "requires" in r.attributes:
+                for req in r.attributes["requires"]:
+                    req_id = Id.parse_id(req)
+                    provides_tree[req_id.resource_str()].append(r.resource_id)
+        # Find skipped for undeployables
+        work = list(undeployable_ids)
+        skippeable: Set[ResourceIdStr] = set()
+        while len(work) > 0:
+            current = work.pop()
+            if current in skippeable:
+                continue
+            skippeable.add(current)
+            work.extend(provides_tree[current])
+        return list(skippeable - set(undeployable_ids))
+
+    async def _put_version(
+        self,
+        env: data.Environment,
+        version: int,
+        rid_to_resource: Dict[ResourceIdStr, data.Resource],
+        unknowns: abc.Sequence[data.UnknownParameter],
+        version_info: Optional[JsonType] = None,
+        resource_sets: Optional[Dict[ResourceIdStr, Optional[str]]] = None,
+        partial_base_version: Optional[int] = None,
+        removed_resource_sets: Optional[List[str]] = None,
+        *,
+        connection: asyncpg.connection.Connection,
+    ) -> None:
+        is_partial_update = partial_base_version is not None
+
+        if resource_sets is None:
+            resource_sets = {}
+
+        if removed_resource_sets is None:
```

Comment:
If this is set, it should be disjunct from `resource_sets`. I don't think there's a check for that anywhere.

Reply:
There is one, but it's [here](https://github.com/inmanta/inmanta-core/blob/76b5e5eb149f82ca3ad156febf502b90216c4e52/src/inmanta/server/services/orchestrationservice.py#L816). Should I copy that check to the `_put_version()` method?

Reply:
For me it's fine to have preconditions here. I think we have to if we want to avoid one monolithic code block. But I do think they should be documented. But I left another comment stating just that so you can resolve this conversation as a subset of the other one.